proto/provenance/exchange/v1/orders.proto

```diff
@@ -0,0 +1,72 @@
+syntax = "proto3";
+package provenance.exchange.v1;
+
+option go_package = "github.com/provenance-io/provenance/x/exchange";
+
+option java_package        = "io.provenance.exchange.v1";
+option java_multiple_files = true;
+
+import "cosmos/base/v1beta1/coin.proto";
+import "cosmos_proto/cosmos.proto";
+import "gogoproto/gogo.proto";
+
+// Order associates an order id with one of the order types.
+message Order {
+  option (gogoproto.goproto_getters) = false;
+
+  // order_id is the numerical identifier for this order.
+  uint64 order_id = 1;
+  // order is the specifics of this order.
+  oneof order {
```

Comment:
I wonder if we should try to stick with Any based on this documentation from cosmos-sdk.

https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-019-protobuf-state-encoding.md#usage-of-any-to-encode-interfaces

Reply:
My problem with using an `any` here is that a codec is then required in order to deal with an `Order` object. Since the codec is an app-level thing that's only available in the keepers, it's significantly more difficult to keep all this logic in this package. Plus, I'm not trying to have this hold an interface that any message can then implement, I only want these two specific types possible in here.

Further, this `Order` message is only used in query results and genesis. Its not used in any tx, signing, or state encoding. In genesis, they'll be json which will just omit the unused type(s). In query results, the chain just uses standard protobuf encoding for the result and clients can use standard protobuf decoding. It doesn't suffer from the same problems discussed in that link (and the issue it references) because all of the possible types are known ahead of time (right now). This isn't designed to be extensible outside of this module. I.e., it's not like a `MsgSubmitProposal` where the content is expected to be defined in a different module/repo/organization.

All that being said, after looking at the `UnpackInterfaces` thing, I might be overblowing the codec complications. And a `oneof` field has a forward compatibility consideration if we ever want to add a type. I.e. If a type is added to this `oneof`, any client using the old proto won't be able to handle results containing the new type. But as an `Any`, it should still be able to decode the content, (just maybe not know what to do with it).

Let me think on this and look into the amount of work that'd be needed to change it.

Reply:
the forward compatibility piece seems like a big deal.. I am suspecting that we already have new order types in the pipeline for the peer to peer use cases...